src/node_api.cc

```diff
@@ -435,6 +428,78 @@ class ThreadSafeFunction : public node::AsyncResource {
   bool handles_closing;
 };
 
+class AsyncContext {
+ public:
+  AsyncContext(node_napi_env env,
+               v8::Local<v8::Object> resource_object,
+               const v8::Local<v8::String> resource_name)
+      : env_(env) {
+    async_id_ = node_env()->new_async_id();
+    trigger_async_id_ = node_env()->get_default_trigger_async_id();
+    resource_.Reset(node_env()->isolate(), resource_object);
+    resource_.SetWeak(this,
```

Comment:
How is it ensured that `resource` is not GCed to early if there is not strong reference? If I compare this with `class node::AsyncResource` the `setWeak()` is not called there.

Reply:
This is required if we don't make a breaking change. See https://github.com/nodejs/node/blob/master/test/node-api/test_make_callback/test-async-hooks-gcable.js , previously the resource object is weak referenced by `napi_async_init`.

For the issue that the resource object is already gc-ed, I updated the PR to return a `napi_invalid_arg` for the condition. I'm not sure if a new n-api status will be better suited in the case.

Reply:
ok, the chicken/egg problem.
test-async-hooks-gcable verifies that resource can be collected and the finalizer deletes the `async_context`. Using a strong reference breaks this as it requires the user to call `napi_async_destroy()` instead of letting finalizer do this.

But this change breaks applications where user code has no reference to the resource. Consider someone is calling
`napi_async_init(env, NULL, name, context)` which creates a resource on the fly [here ](https://github.com/nodejs/node/blob/1d8ecd8cfec2a5340028ac8be9f24024c4f9a810/src/node_api.cc#L684) which can be collected once the function returns.
A later call to `napi_make_callback(env, context, recv, func, argc, argv, &result)` will result in an error if the resource was collected which did not happen before.

Reply:
I think the fact that we have a weak reference is a bug, and fixing it by making the reference strong would not have downsides AFAICT, because
- it would remove the need to return a new error
- it is what Node.js core does with async contexts,
- add-on authors already have to call `napi_async_destroy` in order to avoid memory leaks, so we would still have an opportunity to destroy the strong reference along with the rest of the `napi_async_context`,
- add-on authors
- the test (https://github.com/nodejs/node/blob/master/test/node-api/test_make_callback/test-async-hooks-gcable.js) merely tests that `napi_async_destroy` can be called during a garbage collection run, not necessarily the gc run of the context itself. So, the test could be changed to just `napi_create_object(env, &something); napi_add_finalizer(env, something, (void*)async_context, AsyncDestroyCb, null, null);` instead of attaching the finalizer to `context`. @addaleax please correct me if I'm wrong!

Reply:
> But this change breaks applications where user code has no reference to the resource. Consider someone is calling

The behavior is not changed in the PR. So this PR will not break existing applications.

For the behavior of the `napi_async_init`, `napi_async_init` has to be paired with `napi_async_destroy`. Without `napi_async_destroy` the side effects (memory leaks) are implementation details.

The issue here is the timing that should the resource object be able to be GC-ed before the `napi_async_destroy`.

The current weak referencing behavior is suggesting that the resource object can be GC-ed before the `napi_async_destroy`(and emitting destroy hooks), which is somewhat aligned with the JavaScript AsyncResource API (default behavior). (And in the test case, the `napi_async_destroy` is called on the finalizer of the resource object)

Reference: https://github.com/nodejs/node/pull/27255

However, since the async context created by `napi_async_init` is referenced in C/C++ addons, the context has to be destroyed with `napi_async_destroy` and will not be automatically GC-ed. The async context can still be used with `napi_make_callback`, in which the async hooks trying to set the current execution resource to the creating resource object to get correct async context propagation like in AsyncLocalStorage.

As the timing of garbage collection is not quite reliable in the user code, `napi_async_context` related API will be not determinist in the case of weak referencing the resource object. So I'd propose to change the behavior to strong reference the resource object to make the behavior of `napi_async_context` related API determine-able and robust to users. /cc @nodejs/n-api

Reply:
> and fixing it by making the reference strong would not have downsides AFAICT, because

It can still be a bit of a memory leak footgun if `napi_async_destroy` itself is triggered by a weak callback :)

Reply:
@legendecas I will push a test passing with current node but failing with the PR in the current state later to ensure we discuss the same.

Reply:
see 4b8fcc0f440998fe7c4ca6e8996a5f9873ebb222 for a (hacky) testcase which passes with node master but fails with this PR included.
I would not create such code but it is valid NAPI usage to my understanding.

Reply:
@Flarna thanks for your case, it is very helpful. It seems to me that the issue is still around the weak reference on the resource object. In the case, `napi_async_init` is creating a local object to be the default value of the resource object of the created `napi_async_context`, and `napi_async_init` doesn't return any reference back to caller other than the `napi_async_context`. 

I believe this case can be addressed by making the reference a strong one, too. In this way, the `async_hooks.executionResource()` can maintain a correct context for `AsyncLocalStorage` in the case of `napi_async_init` with NULL resource object.

Reply:
Please note that the problem is not limited to using `NULL` in `napi_async_init()` as it happens also if the test is slightly modified like this:
```
static napi_value CreateContext(napi_env env, napi_callback_info info) {
  napi_value resource_name;
  NAPI_CALL(env, napi_create_string_utf8(
      env, "XXX", NAPI_AUTO_LENGTH, &resource_name));

  napi_value object;
  NAPI_CALL(env, napi_create_object(env, &object));
  NAPI_CALL(env, napi_async_init(env, object, resource_name, &sContext));
}
```
The problem is that till now there was no requirement to usercode to keep the resource alive until the end of the usecase.